pkg/token/serviceprincipaltoken.go

```diff
@@ -88,23 +112,40 @@ func (p *servicePrincipalToken) Token() (adal.Token, error) {
 			return emptyToken, fmt.Errorf("failed to decode pkcs12 certificate while creating spt: %w", err)
 		}
 
-		spt, err = adal.NewServicePrincipalTokenFromCertificate(
-			p.oAuthConfig,
+		cred, err := azidentity.NewClientCertificateCredential(
+			p.tenantID,
 			p.clientID,
-			cert,
+			[]*x509.Certificate{cert},
 			rsaPrivateKey,
-			p.resourceID,
-			callback)
+			clientOptions,
+		)
 		if err != nil {
-			return emptyToken, fmt.Errorf("failed to create service principal token using cert: %s", err)
+			return emptyToken, fmt.Errorf("unable to create credential. Received: %v", err)
 		}
+		spnAccessToken, err = cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{p.resourceID + "/.default"}})
+		if err != nil {
+			return emptyToken, fmt.Errorf("expected an empty error but received: %v", err)
+		}
+
+	} else {
+		return emptyToken, errors.New("service principal token requires either client secret or certificate")
 	}
 
-	err = spt.Refresh()
-	if err != nil {
-		return emptyToken, err
+	if spnAccessToken.Token == "" {
+		return emptyToken, errors.New("did not receive a token")
```

Comment:
let's state the fact in the error message such as `unexpectedly got empty access token`

Reply:
Fixed